RFC/0002-signed-envelopes.md

```diff
@@ -0,0 +1,108 @@
+# RFC 0002 - Signed Envelopes
+
+- Start Date: 2019-10-21
+- Related RFC: [0003 Address Records][addr-records-rfc]
+
+## Abstract
+
+This RFC proposes a "signed envelope" structure that contains an arbitray byte
+string payload, a signature of the payload, and the public key that can be used
+to verify the signature.
+
+This was spun out of an earlier draft of the [address records
+RFC][addr-records-rfc], since it's generically useful.
+
+## Problem Statement
+
+Sometimes we'd like to store some data in a public location (e.g. a DHT, etc),
+or make use of potentially untrustworthy intermediaries to relay information. It
+would be nice to have an all-purpose data container that includes a signature of
+the data, so we can verify that the data came from a specific peer and that it hasn't
+been tampered with.
+
+## Domain Separation
+
+Signatures can be used for a variety of purposes, and a signature made for a
+specific purpose MUST NOT be considered valid for a different purpose.
+
+Without this property, an attacker could convince a peer to sign a payload in
+one context and present it as valid in another, for example, presenting a signed
+address record as a pubsub message.
+
+We separate signatures into "domains" by prefixing the data to be signed with a
+string unique to each domain. This string is not contained within the payload or
+the outer envelope structure. Instead, each libp2p subsystem that makes use of
+signed envelopes will provide their own domain string when constructing the
+envelope, and again when validating the envelope. If the domain string used to
+validate is different from the one used to sign, the signature validation will
+fail.
+
+Domain strings may be any valid UTF-8 string, but should be fairly short and
+descriptive of their use case, for example `"libp2p-routing-record"`.
+
+## Type Hinting
+
+The envelope record can contain an arbitrary byte string payload, which will
+need to be interpreted in the context of a specific use case. To assist in
+"hydrating" the payload into an appropriate domain object, we include a "type
+hint" field. The type hint consists of a [multicodec][multicodec] code,
+optionally followed by an arbitrary byte sequence.
+
+This allows very compact type hints that contain just a multicodec, as well as
+"path" multicodecs of the form `/some/thing`, using the ["namespace"
+multicodec](https://github.com/multiformats/multicodec/blob/master/table.csv#L23),
+whose binary value is equivalent to the UTF-8 `/` character.
+
+## Wire Format
+
+Since we already have a [protobuf definition for public keys][peer-id-spec], we
+can use protobuf for this as well and easily embed the key in the envelope:
+
+
+```protobuf
+message SignedEnvelope {
+  PublicKey publicKey = 1; // see peer id spec for definition
+  bytes typeHint = 2;      // type hint
```

Comment:
shouldn't this be a string?

Reply:
Few points:

* Are we intending to use multicodecs? In that case it should be a bytes.

* The term "type hints" tends to be a term used in reflection/dynamic typing. In this case I would just call this field `payload_class`, or `payload_type`, and the actual content `payload`.

* Convention is to use snake case for protobufs, e.g. from the Kubernetes project: https://github.com/kubernetes/kubernetes/blob/master/third_party/protobuf/google/protobuf/descriptor.proto.

Reply:
+1 on the field naming - I like `payload_type` better than "type hint" & it looks like snake case is the protobuf way.

I did make the type field `bytes` based on @Stebalien's suggestion to use multicodec-prefixed byte strings. Question about that: do we have a go library that validates / parses multicodecs? I found go-multicodec-packed, but it was deprecated a while ago.

Reply:
Not sure if this is what you're looking for, but go-cid has references to the codecs in the multicodec table at https://github.com/ipfs/go-cid/blob/9bb7ea69202c6c9553479eb355ab8a8a97d43a2e/_rsrch/cidiface/enums.go